pkg/util/cidr/cidr.go

```diff
@@ -0,0 +1,328 @@
+		if tickDuration == 0 {
+			tickDuration = time.Hour * 24 * 365
+		}
+		return tickDuration
+	}
+	if err := stopper.RunAsyncTask(ctx, "cidr-refresh", func(ctx context.Context) {
+		ticker := time.NewTicker(getTickDuration())
+		defer ticker.Stop()
+		for {
+			select {
+			case <-stopper.ShouldQuiesce():
+				return
+			case <-ticker.C:
+				c.refresh(ctx)
+			case <-intervalChangedChan:
+				ticker.Reset(getTickDuration())
+			}
+		}
+	}); err != nil {
+		log.Fatalf(ctx, "unable to start CIDR lookup refresh task: %v", err)
+	}
+	return c
+}
+
+// hexString returns a hex string representation of an IP address. The length of
+// the string will always be twice the length of the input.
+func hexString(b []byte) string {
+	const hexDigit = "0123456789abcdef"
+	s := make([]byte, len(b)*2)
+	for i, tn := range b {
+		s[i*2], s[i*2+1] = hexDigit[tn>>4], hexDigit[tn&0xf]
+	}
+	return string(s)
+}
+
+func (c *Lookup) refresh(ctx context.Context) {
+	// Check if the URL is updated
+	if c.isUpdated(ctx, cidrMappingUrl.Get(c.st)) {
+		// Set the URL
+		url := cidrMappingUrl.Get(c.st)
+		if err := c.setURL(ctx, url); err != nil {
+			log.Errorf(ctx, "error setting CIDR URL to '%s': %v", url, err)
+		}
+	}
+}
+
+func (c *Lookup) isUpdated(ctx context.Context, rawURL string) bool {
+	// Check if the URL is a local file
+	if strings.HasPrefix(rawURL, "file://") {
+		// Extract the file path from the URL
+		filePath := strings.TrimPrefix(rawURL, "file://")
+
+		// Get the file information
+		fileInfo, err := os.Stat(filePath)
+		if err != nil {
+			return false
+		}
+
+		// Compare the modification time of the file with lastUpdate
+		if fileInfo.ModTime().After(c.lastUpdate) {
+			c.lastUpdate = fileInfo.ModTime()
+			return true
+		}
+	} else if strings.HasPrefix(rawURL, "http://") || strings.HasPrefix(rawURL, "https://") {
+		// Create an HTTP client
+		client := &http.Client{}
+
+		// Send a HEAD request to the URL
+		resp, err := client.Head(rawURL)
+		if err != nil {
+			return false
+		}
+		defer resp.Body.Close()
+
+		// Get the Last-Modified header from the response
+		lastModified := resp.Header.Get("Last-Modified")
+		if lastModified == "" {
+			return false
+		}
+
+		// Parse the Last-Modified header
+		modTime, err := http.ParseTime(lastModified)
+		if err != nil {
+			return false
+		}
+
+		// Compare the modification time with lastUpdate
+		if modTime.After(c.lastUpdate) {
+			c.lastUpdate = modTime
+			return true
+		}
+	} else if rawURL == "" {
+		if c.lastUpdate.IsZero() {
+			c.lastUpdate = timeutil.Now()
+			return true
+		}
+	}
+
+	return false
+}
+
+// SetURL sets the URL for the CIDR lookup. The URL can be a local file or an
+// HTTP URL. Once the file is loaded it is optionally transformed and then
+// converted into CIDR records and loaded.
+func (c *Lookup) setURL(ctx context.Context, rawURL string) error {
+	contents := []byte("[]")
+	var err error
+	// Check if the URL is a local file
+	if strings.HasPrefix(rawURL, "file://") {
```

Comment:
This is a valid point and it would definitely be a concern for longer lived connections like gRPC. For the primary connections it covers, I'm not as concerned. The SQL connections typically have a lifetime measured in minutes and the backup connection is only for the life of the backup operations. Changefeeds may have long-lived connections to Kafka, so it could matter there, but I'm not sure....

Currently the Start function blocks on initial refresh. I'm not sure if that is ideal, and it would be possible to move this onto the cidr-refresh goroutine fairly easily as well. This might be what happens in practice anyway since the inital call will not have the mapping URL set.

Reply:
>The SQL connections typically have a lifetime measured in minutes and the backup connection is only for the life of the backup operations

This isn't an accurate assumption. It's fairly common for customers to use connection pools with no max connection lifetime. On Serverless, since pods are so volatile, it's likely that connection lifetime is somewhat close to server lifetime.

> Currently the Start function blocks on initial refresh.

I think the map has to be a soft dependency. We can't allow an S3/GCS outage to prevent db nodes from coming online.

Reply:
I agree about the soft dependency, I changed the reload code to only load from the async thread, so it never blocks startup. 

In terms of the long lived connections, as you mentioned, they are somewhat problematic in the sense that they won't pick up any cidr changes. I didn't add this yet, but your suggestion definitely seems doable. It also helps remove any bootstrap issues if there are connections created prior to initialization.

Reply:
This ended up being a bit more complicated than I had originally hoped, but I think the approach I took will work. There are two concerns I wanted to deal with
1) There may be a large number of connections, and when the mapping changes, I didn't want to have thousands of callbacks fire when the mapping changed. 
2) Given a child metric name, it can only be registered once. This means it needs to be centralized in a single map on the server and can't be done "per-connection". 

I took a couple stabs at doing this, but ended up with the current approach. I am definitely open to feedback or other alternatives on how to do this a little better. 

Note that I also updated the CIDR commit to add the SetOnChange method.